x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/user/SystemUser.java

```diff
@@ -44,10 +46,15 @@ public static boolean is(User user) {
         return INSTANCE.equals(user);
     }
 
-    public static boolean is(String principal) {
+    public static boolean deserializesFrom(String principal) {
         return NAME.equals(principal);
     }
 
+    public static void write(StreamOutput output) throws IOException {
```

Comment:
Example change for `SystemUser`. Will apply to all other internal users if this looks good.

Reply:
I'd have this method with signature `public void writeTo(StreamOutput)` which feels more consistent with other places. Static `write` method often takes an extra argument which is the thing to be written.

Reply:
Making it not static would require using `INSTANCE`, which is of type `User` and so won't have the custom `writeTo` method exposed. 

I'll go back to inlining this in Authentication, along with inlining `is`. Will then revisit in a small follow up PR, if we decide that encapsulating `NAME` is worth it after all.

Reply:
We'd have to use it here:

https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java#L978-L979

i.e., 

```
if (SystemUser.is(user)) {
    SystemUser.INSTANCE.writeTo(output);
}
```

But `INSTANCE` is a `User` and does not expose `writeTo`

Reply:
> But `INSTANCE` is a `User` and does not expose `writeTo`

Right. That can be changed too now that User class is simplified (#86246). Previously, user can have inner user which causes recursion problem for a regular `writeTo` method. This is no longer the case. Also the `SystemUser.is(User)` method isn't necessary either. If we mark `SystemUser` as `final`, the `is` method can be replaced by `instanceOf` or maybe even object identity comparison. 

All is to say there are quite a few thing we can do to improve internal users. But again, I think we can revisit later.